rfcs/037-api-faceting-principles/README.md

```diff
@@ -0,0 +1,146 @@
+
+```
+{
+  ...,
+  "aggregations": {
+    "a.b.c": {
+       "buckets": [
+         ...
+       ]
+    }
+  }
+}
+```
+
+
+
+**4. Aggregation buckets contain a `data` field of the same type as the aggregated property's parent entity**
+
+That is to say, while we aggregate on a specific string field (for example, an ID or a label), we want to return the full entity that contains the field. This is easiest to explain by example. Given the following display documents:
+
+```json
+{
+  "a": {
+    "b": [
+      {
+        "id": "id1",
+        "label": "Thing 1"
+      },
+      {
+        "id": "id2",
+        "label": "The 2nd thing"
+      }
+    ]
+  }
+}
+```
+
+Then if we aggregate on the labels like this:
+
+```
+http://host.name/path/docs?aggregations=a.b.label
+```
+
+Then our response buckets will look something like this:
+
+```
+{
+  "data": {
+    "id": "id1",
+    "label": "Thing 1"
+  },
+  "count": 1234
+}
+```
+
+**5. When a filter and its paired aggregation are both applied, that aggregation's buckets are not filtered**
+
+Conversely, filters do apply to the buckets of all aggregations other than the paired aggregation. This initially confusing requirement is necessary because - for mutually exclusive values - application of the filter to the aggregation buckets will remove all but the selected bucket, thus removing the ability of the interface to show other options for the given filter. This is explained further in the [Elasticsearch documentation for `post_filter`](https://www.elastic.co/guide/en/elasticsearch/reference/current/filter-search-results.html#post-filter).
+
+**6. When a filter and its paired aggregation are both applied, the bucket corresponding to the filtered value is always present**
+
+Explicitly: even if other filters or queries are present which cause that bucket to be empty (ie, it has a count of 0), it still appears in the aggregation. This is necessary so that the interface for the filter can still be rendered.
+
+**7. Aggregations on fields contained in sum types return buckets of the type's components**
+
+In other words - there can be a discriminator present on objects meaning that aggregations on identical string properties of those objects return separate objects for each type. For example, given the following display documents:
+
+```json
+{
+  "a": {
+    "b": [
+      {
+        "label": "A thing",
+        "type": "TypeOne"
+      },
+      {
+        "label": "A thing",
+        "type": "TypeTwo"
+      }
+    ]
+  }
+}
+```
+
+Then an aggregation `a.b.label` would return separate buckets for each of the objects in `b`, even though their labels (the property being aggregated) are identical, because of the presence of the discriminator field `type`.
+
+## Open questions
+
+- Are these rules sufficient to tell us what we expect regarding empty buckets? Namely, that they should only be present when necessary to satisfy principle (6).
```

Comment:
Yes. I think we only want empty buckets when we need to show that something has been applied as a filter, not when applied filters cause buckets that would otherwise be there to be empty.

Reply:
I think this might require a bit of thought to implement but will mean we can get rid of the hated `min_doc_count: 0`, which is an added bonus